src/lib/utils/index.ts

```diff
@@ -326,6 +326,23 @@ export type ContextualError<Context extends Record<string, unknown> = {}> = Erro
   context: Context
 }
 
+export interface Exception<T extends string, C extends SomeRecord> extends Error {
+  type: T
+  context: C
+}
+
+export function execeptionType<Type extends string, Context extends SomeRecord>(
+  type: Type,
+  template: (ctx: Context) => string
+) {
+  return (ctx: Context) => {
+    const e = new Error(template(ctx)) as Exception<Type, Context>
+    e.type = type
+    e.context = ctx
+    return e
+  }
+}
```

Comment:
I am not satisfied with the usability but it seems to get the job done which is a two step processes where we:

1. define error types
2. create instances of those error types

I am open to using classes but didn't see how that would help.

I hope we can make this better in the future.

Reply:
I find it interesting as a first iteration and certainly not bad. Let's simply see how that works in practice for now and iterate if it turns out to not work well!